### PR TITLE
Add Flower ip_proto,  dst_port and src_port attributes support

### PR DESCRIPTION
**Pull request by e0ne**

### CODE REVIEW

File: filter_test.go [hunk 1]
```diff
@@ -1524,6 +1527,7 @@ func TestFilterFlowerAddDel(t *testing.T) {
 		EncSrcIPMask:  testMask,
 		EncDestPort:   8472,
 		EncKeyId:      1234,
+		IPProto:       ipproto,
```
Comment (aboch) [inline filter_test.go:1530]: I do not see this function checking this parameter once it reads the added filter from kernel.

File: nl/nl_linux.go [hunk 2]
```diff
@@ -789,3 +789,36 @@ func (sh *SocketHandle) Close() {
 		sh.Socket.Close()
 	}
 }
+
+// IPProto represents Flower ip_proto attribute
```
Comment (adrianchiris) [inline nl/nl_linux.go:793]: im not sure if these bits should be here as they are specific to tc filter.

maybe move to `tc_linux.go` ?

File: filter_linux.go [hunk 3]
```diff
@@ -129,6 +132,30 @@ func (filter *Flower) encode(parent *nl.RtAttr) error {
...
+			case nl.IPPROTO_UDP:
+				parent.AddRtAttr(nl.TCA_FLOWER_KEY_UDP_SRC, htons(filter.SrcPort))
+			case nl.IPPROTO_SCTP:
+				parent.AddRtAttr(nl.TCA_FLOWER_KEY_SCTP_SRC, htons(filter.SrcPort))
+			}
+		}
+		if filter.DstPort != 0 {
+			switch ipproto {
+			case nl.IPPROTO_TCP:
+				parent.AddRtAttr(nl.TCA_FLOWER_KEY_TCP_DST, htons(filter.SrcPort))
```
Comment (adrianchiris) [inline filter_linux.go:151]: that needs to be filter.DstPort

File: filter_linux.go [hunk 4]
```diff
@@ -162,6 +189,14 @@ func (filter *Flower) decode(data []syscall.NetlinkRouteAttr) error {
...
 		case nl.TCA_FLOWER_KEY_ENC_KEY_ID:
 			filter.EncKeyId = ntohl(datum.Value)
+		case nl.TCA_FLOWER_KEY_IP_PROTO:
+			val := new(nl.IPProto)
+			*val = nl.IPProto(datum.Value[0])
+			filter.IPProto = val
+		case nl.TCA_FLOWER_KEY_TCP_SRC, nl.TCA_FLOWER_KEY_UDP_SRC, nl.TCA_FLOWER_KEY_SCTP_SRC:
+			filter.SrcPort = ntohs(datum.Value)
+		case nl.TCA_FLOWER_KEY_TCP_DST, nl.TCA_FLOWER_KEY_UDP_DST, nl.TCA_FLOWER_KEY_SCTP_DST:
+			filter.SrcPort = ntohs(datum.Value)
```
Comment (adrianchiris) [inline filter_linux.go:199]: that needs to be filter.DstPort

File: filter_test.go [hunk 5]
```diff
@@ -1586,6 +1590,9 @@ func TestFilterFlowerAddDel(t *testing.T) {
 	if filter.EncDestPort != flower.EncDestPort {
 		t.Fatalf("Flower EncDestPort doesn't match")
 	}
+	if flower.IPProto != nil && *filter.IPProto != *flower.IPProto {
+		t.Fatalf("Flower EncDestPort doesn't match")
```
Comment (adrianchiris) [inline filter_test.go:1594]: wrong message

File: filter_test.go [hunk 6]
```diff
@@ -1506,6 +1506,9 @@ func TestFilterFlowerAddDel(t *testing.T) {
 
 	testMask := net.CIDRMask(24, 32)
 
+	ipproto := new(nl.IPProto)
+	*ipproto = nl.IPPROTO_ICMP
+
```
Comment (adrianchiris) [inline filter_test.go:1511]: @e0ne  could you also add Src and Dst Port to the test?

File: filter_test.go [hunk 7]
```diff
@@ -1586,6 +1590,9 @@ func TestFilterFlowerAddDel(t *testing.T) {
 	if filter.EncDestPort != flower.EncDestPort {
 		t.Fatalf("Flower EncDestPort doesn't match")
 	}
+	if flower.IPProto != nil && *filter.IPProto != *flower.IPProto {
```
Comment (adrianchiris) [inline filter_test.go:1593]: @e0ne i believe we expect flower.IPProto to not be nil  else we should fail.

so this condition should be:

```
if flower.IPProto == nil || *flower.IPProto != filter.IPProto {
   t.Fatalf("Flower IPProto doesn't match")
}
```

File: filter_linux.go [hunk 8]
```diff
@@ -64,6 +64,9 @@ type Flower struct {
 	EncSrcIPMask  net.IPMask
 	EncDestPort   uint16
 	EncKeyId      uint32
+	IPProto       *nl.IPProto
+	DstPort       uint16
```
Comment (adrianchiris) [inline filter_linux.go:68]: For consistency can you name this one `DestPort` ? like `DestIP` and other Dest* attributes in this struct ?

File: filter_test.go (hunk 1 above)
Comment (e0ne) [reply filter_test.go:1530]: The check is added to #1594

File: filter_test.go [hunk 9]
```diff
@@ -1586,6 +1593,15 @@ func TestFilterFlowerAddDel(t *testing.T) {
 	if filter.EncDestPort != flower.EncDestPort {
 		t.Fatalf("Flower EncDestPort doesn't match")
 	}
+	if flower.IPProto != nil || *filter.IPProto != *flower.IPProto {
```
Comment (adrianchiris) [inline filter_test.go:1596]: should be : `flower.IPProto == nil || *filter.IPProto != *flower.IPProto`

**Review by adrianchiris — APPROVED** [review]

**Comment by e0ne** [comment]
@aboch @vishvananda  could you please take a look on this PR once you have a time?

**Comment by e0ne** [comment]
just a kindly ask to @aboch @vishvananda for review

**Comment by adrianchiris** [comment]
Thats the last bit required for: https://github.com/k8snetworkplumbingwg/multi-networkpolicy-tc/pull/47

@aboch @vishvananda  appreciate your review for this one :)

**Comment by aboch** [comment]
Thanks @e0ne 

The two commits address distinct things, I will merge as is.

**Pull request merged**